File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/tax/usecases/ReturnsTaxRate.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.tax.usecases;

import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.VoidVerification;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.tax.usecases.base.BaseTaxUseCase;
import com.mycompany.myshop.testkit.driver.port.external.tax.TaxDriver;
import com.mycompany.myshop.testkit.driver.port.external.tax.dtos.ReturnsTaxRateRequest;

public class ReturnsTaxRate extends BaseTaxUseCase<Void, VoidVerification> {
    private String countryAlias;
    private String taxRate;

    public ReturnsTaxRate(TaxDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public ReturnsTaxRate country(String countryAlias) {
        this.countryAlias = countryAlias;
        return this;
    }

    public ReturnsTaxRate taxRate(String taxRate) {
        this.taxRate = taxRate;
        return this;
    }

    public ReturnsTaxRate taxRate(double taxRate) {
        return taxRate(Converter.fromDouble(taxRate));
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var country = context.getParamValueOrLiteral(countryAlias);

        var request = ReturnsTaxRateRequest.builder()
                .country(country)
                .taxRate(taxRate)
                .build();

        var result = driver.returnsTaxRate(request);

        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/tax/usecases/ReturnsTaxRate.ts
import type { TaxDriver } from '../../../../../../driver/port/external/tax/tax-driver.js';
import type { ReturnsTaxRateRequest } from '../../../../../../driver/port/external/tax/dtos/ReturnsTaxRateRequest.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import { VoidVerification } from '../../../../shared/void-verification.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseTaxUseCase } from './base/BaseTaxUseCase.js';

export class ReturnsTaxRate extends BaseTaxUseCase<void, VoidVerification> {
  private _countryAlias?: string;
  private _taxRate: number | string = '';

  constructor(driver: TaxDriver, context: UseCaseContext) {
    super(driver, context);
  }

  country(countryAlias: string): this {
    this._countryAlias = countryAlias;
    return this;
  }

  taxRate(taxRate: number | string): this {
    this._taxRate = taxRate;
    return this;
  }

  async execute(): Promise<UseCaseResult<void, VoidVerification>> {
    const country = this.context.getParamValueOrLiteral(this._countryAlias);

    const request: ReturnsTaxRateRequest = {
      country: country ?? '',
      taxRate: String(this._taxRate),
    };

    const driverResult = await this.driver.returnsTaxRate(request);
    const result: Result<void, SystemError> = driverResult.success
      ? success(undefined)
      : failure({ message: driverResult.error.message ?? 'Tax error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
